### ISSUE TITLE
[AIR] Syncing when upload_url="s3://<my_s3_bucket>?<endpoint_and_other_argument_overrides>"

### ISSUE THREAD

**Issue by tbukic** [ray-project/ray#29845]
### What happened + What you expected to happen

When using non-AWS S3 API supporting storage, it seems that to get past  `_validate_upload_dir(sync_config)` in [L454 tune.py](https://github.com/ray-project/ray/blob/master/python/ray/tune/tune.py#L454) an `upload_dir` has to be provided in the shape of `s3://<my_s3_bucket>?endpoint_override=<endpoint_url[&possible_other_args_like_region]>`.

In that case, syncer gets an upload url in form `s3://<my_s3_bucket>?endpoint_override=<endpoint_url>/<ray_defined_upload_path>`. Thus, 'auto'-configured Syncer fails to upload files since PyArrow needs the path to be in the form of `s3://<my_s3_bucket>/<ray_defined_upload_path>?endpoint_override=<endpoint_url>`.

Construction `os.path.join(self.local_dir, self.dir_name)` in [experiment.py](https://github.com/ray-project/ray/blob/master/python/ray/tune/experiment/experiment.py) seems to be responsible for this. This can be fixed with custom syncer which fixes URIs, but Ray should take care of passing correct URI unless another way to provide `endpoint_override` exists. As far as I know no other way is specified either in Ray's documentation or in the documentation of Ray's dependencies.

Example of almost-working syncer in the case when `endpoint_override` is defined in `upload_dir` variable.

```
def fix_uri(uri: str) -> str:
    bucket, rest = uri.split('?', 1)
    querry, path = rest.split('/', 1)
    uri = f'{os.path.join(bucket, path)}/?{querry}'
    return uri


class FixedSyncer(_DefaultSyncer):
    def _sync_up_command(
        self, local_path: str, uri: str, exclude: List | None = None
    ) -> Tuple[Callable, Dict]:
        uri = fix_uri(uri)
        return super()._sync_up_command(local_path, uri, exclude)

    def _sync_down_command(self, uri: str, local_path: str) -> Tuple[Callable, Dict]:
        uri = fix_uri(uri)
        return super()._sync_down_command(uri, local_path)

    def _delete_command(self, uri: str) -> Tuple[Callable, Dict]:
        uri = fix_uri(uri)
        return super()._delete_command(uri)
```

Second problem is that, when uploading checkpoints, sign '=' in tune-defined filename doesn't get accepted by either PyArrow or my S3 provider (OVH cloud). I'm not sure if this is OVH specific problem and how serious Ray think this issue is in syncing.

### Versions / Dependencies

ray 2.0.1

### Reproduction script

Please set the breakpoint at the first line of `_sync_up_command` in `FixedSyncer` (or `_DefaultSyncer` if you wish to drop `syncer` argument from the example).
To get to this point you'll need either an existing custom non-aws s3 storage and correct credentials, or comment out `_validate_upload_dir(sync_config)` in [tune.py](https://github.com/ray-project/ray/blob/master/python/ray/tune/tune.py#L454) because bad path can be seen even before needing to interact with the actual storage.

```
tune.Tuner(
        "PPO",
        run_config=air.RunConfig(
            stop={"episode_reward_mean": 200},
            sync_config=SyncConfig(
                syncer=FixedSyncer(sync_period=60),
                upload_dir=your_upload_dir_**containing_endpoint_override**
            ),
        ),
        tune_config=tune.TuneConfig(
            metric="episode_reward_mean",
            mode="max",
        ),
        param_space={
            "env": "CartPole-v1",
            "framework": "torch",
            "num_gpus": 0,
            "num_workers": 1,
            "lr": tune.grid_search([0.01, 0.001])
        },
).fit()
```

Replicating problems created by '=' sign in S3 filename for me includes creating the breakpoint at  [remote_storage.py#L209](https://github.com/ray-project/ray/blob/master/python/ray/air/_internal/remote_storage.py#L209)
and executing edited line in debugger terminal: `pyarrow.fs.copy_files(local_path, bucket_path.replace('=', '-'), destination_filesystem=fs)` (passes) before/instead executing the original line (fails). For this part OVH s3 bucket is necessary, and I'm unsure about the other providers.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by tbukic** [comment]
One additional detail:

To be able to get the whole `ray_results` synced to my s3 storage, I also needed to encode `full_target_path` in [remote_storage.py#L236](https://github.com/ray-project/ray/blob/master/python/ray/air/_internal/remote_storage.py#L236,), because '=' was preventing lot of files from being synced - including the tensorboard logs. Thus, my current (working) version encodes URIs after 'fixing them' (as shown in the original issue), and than encodes them in all 3 syncer calls + for `full_target_path` in `remote_storage:_upload_to_uri_with_exclude`.

**Comment by tbukic** [comment]
Sorry, this is accidentally marked as completed. Just wanted to add the update. This is still an issue.

**Comment by bveeramani** [comment]
Hey @tbukic, thanks for reporting this issue! The detailed writeup is super helpful.

Just a few questions to confirm my own understanding:
1. You're passing a URI to `upload_dir` that contains URL-encoded arguments, but the program errors because `remote_checkpoint_dir` incorrectly appends the filename after the URL-encoded arguments?
```
    @property
    def remote_checkpoint_dir(self) -> Optional[str]:
        if not self.sync_config.upload_dir or not self.dir_name:
            return None
        # This doesn't work if `upload_dir` has URL-encoded arguments?
        return os.path.join(self.sync_config.upload_dir, self.dir_name)
```
2. Some of your files contain the `=` character, and this causes the syncer to error. Could you tell me more about the error you received in this case?

Also, just out of curiosity, why're you using a non-AWS API over an AWS API?

**Comment by tbukic** [comment]
Hello @bveeramani 

> Just a few questions to confirm my own understanding:
> 
>     1. You're passing a URI to `upload_dir` that contains URL-encoded arguments, but the program errors because `remote_checkpoint_dir` incorrectly appends the filename after the URL-encoded arguments?
> 
> 
> ```
>     @property
>     def remote_checkpoint_dir(self) -> Optional[str]:
>         if not self.sync_config.upload_dir or not self.dir_name:
>             return None
>         # This doesn't work if `upload_dir` has URL-encoded arguments?
>         return os.path.join(self.sync_config.upload_dir, self.dir_name)
> ```

This is true. Code I have in my .venv is a bit older (despite I use the latest released version!) and does not have this functionality extracted to the function, but the logic of appending `ray_results` path to user provided remote path is the same.
 
>     2. Some of your files contain the `=` character, and this causes the syncer to error. Could you tell me more about the error you received in this case?

Indeed. This is the example of file that wouldn't get synced without escaping `=` character: `<my_bucket_name>/PPO/PPO_CartPole-v1_cc07c_00001_1_lr%3D0.0010_2022-11-01_03-17-20/params.json`. Actually, no folder from inside `ray_results/PPO` gets synced.

This is the error which I get when I disable escaping of `=` sign:

> ray::PPO.save() (pid=5263, ip=172.30.212.107, repr=PPO)
>   File "<dir>/.venv/lib/python3.10/site-packages/ray/tune/syncer.py", line 294, in wait_or_retry
>     self.wait()
>   File <dir>/.venv/lib/python3.10/site-packages/ray/tune/syncer.py", line 397, in wait
>     raise TuneError(f"Sync process failed: {e}") from e
> ray.tune.error.TuneError: Sync process failed: When initiating multiple part upload for key '<bucket>/PPO/PPO_CartPole-v1_41820_00000_0_lr=0.0100_2022-11-02_00-23-56/checkpoint_000009/.is_checkpoint' in bucket '<bucket>': AWS Error [code 22]: The request signature we calculated does not match the signature you provided. Check your key and signing method.
> 
> The above exception was the direct cause of the following exception:
> 
> ray::PPO.save() (pid=5263, ip=172.30.212.107, repr=PPO)
>   File "<dir>/.venv/lib/python3.10/site-packages/ray/tune/trainable/trainable.py", line 496, in save
>     self._maybe_save_to_cloud(checkpoint_dir)
>   File "<dir>/.venv/lib/python3.10/site-packages/ray/tune/trainable/trainable.py", line 508, in _maybe_save_to_cloud
>     self.custom_syncer.wait_or_retry()
>   File "<dir>/.venv/lib/python3.10/site-packages/ray/tune/syncer.py", line 305, in wait_or_retry
>     raise TuneError(
> ray.tune.error.TuneError: Failed sync even after 3 retries.

I've figured out manually by creating the break-point just before syncing that `=` is the problem. Without `=`, pyarrow call passes. With `=`, it errors out and the error is not the most informative one.

> Also, just out of curiosity, why're you using a non-AWS API over an AWS API?

Storage provider is one of project requirements.

**Comment by bveeramani** [comment]
@tbukic got it! Thanks for the fast response.

It looks like these issues should be easy to patch. I'll open a PR in the few weeks.

**Comment by bveeramani** [comment]
I've opened a PR to fix the `remote_checkpoint_dir` issue: https://github.com/ray-project/ray/pull/30125.

I don't think we can elegantly handle the "=" issue on our end. I suspect Arrow doesn't URL-encode special characters like "=", and that causes issues for OVH. I checked and it's not a problem for AWS. In any case, I've filed an issue with Arrow: https://issues.apache.org/jira/browse/ARROW-18290.
